### ISSUE TITLE
Intra-endpoint horizontal scaling

### ISSUE THREAD

**Issue by Ralith** [quinn-rs/quinn#1576]
Quinn, like any flexible QUIC implementation, supports horizontal scaling across multiple endpoints through the use of custom connection ID generators and QUIC-aware load-balancing front-ends. However, the requirement for a third party load balancer and custom logic makes this difficult to leverage. For the overwhelmingly common case of applications that fit comfortably on a single host, Quinn should allow a single endpoint to scale the number of concurrent near-linearly with respect to available CPU cores.

In the current architecture, separate connection drivers already allow significant cryptographic work, and all application-layer work, for independent connections to happen in parallel. A bottleneck remains at the endpoint driver, an async task responsible for driving all network I/O and timers. We can do better.

The Linux [`SO_REUSEPORT`](https://lwn.net/Articles/542629/) option will distribute incoming packets on a single port across multiple sockets. Windows "Receive Side Scaling" may be similar. On at least these platforms, we can therefore run up to one endpoint driver per core. Some architectural changes are required to prevent catastrophic contention.

Key `quinn_proto::Endpoint` methods take `&mut self`, preventing meaningful parallelization of datagram processing. We should embrace interior mutability to convert these to `&self` methods which will not contend on the hot path of handling datagrams of established connections. In particular:
- Independent `Connection` states *must* be concurrently mutable. Morally this could be as simple as `RwLock<Slab<Mutex<Connection>>>`.
- Global locking on connection creation/destruction is difficult to avoid, but probably tolerable. See also sharded-slab, a complex but fast storage option.
- An explicit "endpoint view" notion in quinn-proto might be useful to allow lock-free access to per-socket state such as transmit queues without pushing too much more complexity into the `quinn` layer.

SO_REUSEPORT consistently routes packets to sockets based on a 4-tuple hash, so we can rely on connections moving between drivers only in the event of migrations, minimizing contention outside of connection setup/teardown.

Tokio presently lacks a mechanism to ensure certain tasks run on separate threads, which may complicate improving parallelism for high-level users. We could address this by spawning our own threads, or by working with upstream to develop new APIs. It's also possible that simply spawning N drivers and letting work stealing do its thing might work out well enough in practice.

### Unified Drivers

In past discussions this refactoring has often been associated with the unification of endpoint and connection drivers (e.g. #1219). I believe they can be separated, though moving `quinn_proto::Connection` ownership into `quinn_proto::Endpoint` may still be desirable for API simplicity and to reduce perhaps costly inter-thread communications.

Flattening connection tasks complicates timer handling, since we don't want to poll each connection's timer on every endpoint wakeup. My  [`timer-queue`](https://github.com/Ralith/timer-queue) crate provides a solution. Each endpoint driver could maintain timers for connections involved in traffic passing through that driver. By discarding timeouts for a connection that has seen activity more recently on another driver we can ensure timers do not reintroduce contention.

### Future Work

Opportunity also exists for fine-grained parallelism between streams on a single connection, similar in spirit to connections in a single endpoint.

**Comment by Ralith** [comment]
On review, handling multiple connections on a single task might actually be undesirable as it prevents tokio's work-stealing from redistributing work when e.g. a single connection involves disproportionately large amounts of work. Splitting up the endpoint task is still probably valuable, but maybe we don't want to inline connection tasks after all.

Also note tokio's multithreaded runtime relies on a single global epoll loop to drive I/O across all threads, which reportedly allows for more efficient work stealing than epoll-per-thread would. Unclear if this will be a significant bottleneck for Quinn.

**Comment by PureWhiteWu** [comment]
Hi, I've encountered related issue, when handling thousands of connections, quinn server cannot be scaled horizontally.

**Comment by taikulawo** [comment]
kernel use 4-tuple to balance between multiple socket. Can we  create many endpoint to mitigate this issue?

**Comment by Ralith** [comment]
Running independent QUIC server endpoints on the same socket is incompatible with client migration. If you're willing to turn migration off, then that might work. In the long run, Quinn should address this use case directly by allowing parallelization of endpoint-level work. Note that we already parallelize connection-level work.

**Comment by taikulawo** [comment]
I see nginx/envoy use bpf to balance cid to right socket, maybe we can also use.

**Comment by taikulawo** [comment]
one endpoint is our bottleneck, we can't use all cores to accept quic stream. it's more important than connection migration

**Comment by Ralith** [comment]
> I see nginx/envoy use bpf to balance cid to right socket, maybe we can also use.

Yes, if you're comfortable setting up and deploying a packet-level load-balancing front-end, then that's absolutely the best way to go. That sets you up to scale not just across cores but across machines, and doesn't require any changes to Quinn. If you'd like help doing so, please open a new issue; this issue is for discussion of improved scaling of a single endpoint without a load balancer.

**Comment by gretchenfrage** [comment]
I've been thinking about this. Did a bit of experimenting in [my `par` branch](https://github.com/gretchenfrage/quinn/commits/par/).

> Global locking on connection setup/teardown is difficult to avoid, but probably tolerable. See also sharded-slab, a complex but fast storage option.

Maybe we can avoid it? I'd be happy to discuss. There's some pretty radical things I think we could do.

**Comment by Ralith** [comment]
Definitely interested! I'm not *too* concerned about locking on setup/teardown, since many interesting use cases won't be bottlenecked there. I think the first step will be to support receive-side parallelism in the steady state, with a large number of active connections. Once we've got that, further reducing locking certainly won't hurt.

I see you're relying on DashMap to do a lot of the heavy lifting in your branch. I'm a little intimidated by that structure's complexity (similar to sharded-slab, I figure), though I haven't studied it in depth.

**Comment by gretchenfrage** [comment]
> I see you're relying on DashMap to do a lot of the heavy lifting in your branch. I'm a little intimidated by that structure's complexity (similar to sharded-slab, I figure), though I haven't studied it in depth.

From what I've seen, which is limited, dashmap looks to use [pretty simple sharding](https://docs.rs/dashmap/latest/src/dashmap/lib.rs.html#89-93) (it seems like morally it's a `Box<[RwLock<HashMap<K, V>>]>`, rather than attempting to be a lock-free structure, whereas sharded-slab is doing some more [complex](https://docs.rs/sharded-slab/latest/sharded_slab/implementation/index.html) stuff and advertises itself as using lock-free mechanisms. Sharded-slab has [tunable configuration](https://docs.rs/sharded-slab/latest/sharded_slab/trait.Config.html).

I'm sure it's a great project for what it is, but it does worry me that, for example, [sharded-slab `insert` can fail](https://docs.rs/sharded-slab/latest/sharded_slab/struct.Slab.html#method.insert) if "the shard for the current thread is full and no items can be added until some are removed, or the maximum number of shards has been reached" (it was not obvious to me within 5 minutes of looking at that what exactly this means or what the implications of this would be).

The way I see it:

- If we are parallelizing the steady state and serializing connection creation/destruction, we should be able to do something like wrapping the `ConnectionIndex` in a `RwLock` and converting `Slab<ConnectionMeta/IncomingBuffer>` to `RwLock<Slab<Mutex<ConnectionMeta/IncomingBuffer>>>`. Sharded-slab would not optimize this more.
- If we are parallelizing everything, we should be able to do something like changing the `ConnectionIndex`'s `HashMap`s to `DashMap`s, but also, rather than directly elevating our `Slab`s to a concurrent version of them like `sharded_slab::Slab`, avoid that altogether with some mechanism such as mapping the CIDs and FourTuples directly to `Arc<Mutex<ConnectionMeta/IncomingBuffer>>` or something like that.

  Basically, in my opinion, using `Slab`s rather than storing some other sort of value or reference directly is a layer of indirection that simplifies our architecture in the current case, but trying to stick to it in the concurrent case seems like it might be losing the plot--I suspect it would start to lose its simplifying advantage, and introduce synchronization overhead and race condition concerns that could be avoided by doing something else.

**Comment by Ralith** [comment]
> dashmap looks to use [pretty simple sharding](https://docs.rs/dashmap/latest/src/dashmap/lib.rs.html#89-93)

Oh, that's much simpler than I thought it was. Maybe I was confusing it with a different lib, or maybe they redesigned it since I last heard anything. Seems like a great choice, then!

> I'm sure it's a great project for what it is, but it does worry me that, for example, [sharded-slab insert can fail](https://docs.rs/sharded-slab/latest/sharded_slab/struct.Slab.html#method.insert)

Yeah, that's the kind of complexity that unsettles me. I didn't mean to suggest sharded-slab would be a better choice, just that it is an intimidatingly complex parallel data structure with similar capabilities. Happy to drop `Slab` patterns in an intra-endpoint-scalable refactor if that's what makes sense.

**Comment by Ralith** [comment]
@matheus23 notes that https://github.com/ibraheemdev/papaya might be another good parallel hash table.
